dbms/src/Interpreters/JoinV2/SemiJoinProbe.cpp

```diff
@@ -0,0 +1,886 @@
+        // Sanity check
+        RUNTIME_CHECK(ctx.left_semi_match_res.size() == ctx.rows);
+        if (!join->non_equal_conditions.other_eq_cond_from_in_name.empty())
+            RUNTIME_CHECK(ctx.left_semi_match_null_res.size() == ctx.rows);
+    }
+
+    wd.selective_offsets.clear();
+    wd.selective_offsets.reserve(settings.max_block_size);
+
+    size_t left_columns = join->left_sample_block_pruned.columns();
+    size_t right_columns = join->right_sample_block_pruned.columns();
+    if (!wd.result_block)
+    {
+        RUNTIME_CHECK(left_columns + right_columns == join->all_sample_block_pruned.columns());
+        for (size_t i = 0; i < left_columns + right_columns; ++i)
+        {
+            ColumnWithTypeAndName new_column = join->all_sample_block_pruned.safeGetByPosition(i).cloneEmpty();
+            new_column.column->assumeMutable()->reserveAlign(settings.max_block_size, FULL_VECTOR_SIZE_AVX2);
+            wd.result_block.insert(std::move(new_column));
+        }
+    }
+
+    MutableColumns added_columns(right_columns);
+    for (size_t i = 0; i < right_columns; ++i)
+        added_columns[i] = wd.result_block.safeGetByPosition(left_columns + i).column->assumeMutable();
+
+    Stopwatch watch;
+    if (pointer_table.enableProbePrefetch())
+    {
+        probeFillColumnsPrefetch<KeyGetter, kind, has_null_map, has_other_eq_cond_from_in, tagged_pointer>(
+            ctx,
+            wd,
+            added_columns);
+    }
+    else
+    {
+        probeFillColumns<KeyGetter, kind, has_null_map, has_other_eq_cond_from_in, tagged_pointer>(
+            ctx,
+            wd,
+            added_columns);
+    }
+    wd.probe_hash_table_time += watch.elapsedFromLastTime();
+
+    // Move the mutable column pointers back into the wd.result_block, dropping the extra reference (ref_count 2→1).
+    // Alternative: added_columns.clear(); but that is less explicit and may misleadingly imply the columns are discarded.
+    for (size_t i = 0; i < right_columns; ++i)
+        wd.result_block.safeGetByPosition(left_columns + i).column = std::move(added_columns[i]);
+
+    if (!wd.selective_offsets.empty())
+    {
+        for (size_t i = 0; i < left_columns; ++i)
+        {
+            if (!join->left_required_flag_for_other_condition[i])
+                continue;
+            wd.result_block.safeGetByPosition(i).column->assumeMutable()->insertSelectiveFrom(
+                *ctx.block.safeGetByPosition(i).column.get(),
+                wd.selective_offsets);
+        }
+
+        wd.replicate_time += watch.elapsedFromLastTime();
+
+        handleOtherConditions<KeyGetter, kind>(ctx, wd);
+    }
+
+    if (ctx.isProbeFinished())
+    {
+        if constexpr (kind == Semi || kind == Anti)
+            return genResultBlockForSemi(ctx);
+        else
+            return genResultBlockForLeftOuterSemi(ctx, has_other_eq_cond_from_in);
+    }
+
+    return join->output_block_after_finalize;
+}
+
+static constexpr UInt16 INITIAL_PACE = 4;
```

Comment:
the original implementation is 2 for `INITIAL_PACE`, why change it to 4?

Reply:
Because I think the join_v2 implementation needs more context data than the join_v1 one, especially for the prefetch version. Make it a little larger to reduce the saving context data overhead.